<a href="https://colab.research.google.com/github/Vynogradova/homework/blob/main/Homework_9_3_%22_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [ ]:
import pandas as pd

In [6]:
df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/supermarket_sales.csv")

print(type(df))

<class 'pandas.core.frame.DataFrame'>


1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [9]:
top_5_expensive = df.sort_values(by='Unit price', ascending=False).head(5)
print(top_5_expensive)

      Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rati

2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [10]:
electronic_df = df[df['Product line'] == 'Electronic accessories']

max_qty_idx = electronic_df['Quantity'].idxmax()

max_price_idx = electronic_df['Unit price'].idxmax()

print("Рядок з максимальною кількістю:")
print(df.loc[max_qty_idx])

print("\nРядок з максимальною ціною:")
print(df.loc[max_price_idx])

Рядок з максимальною кількістю:
Invoice ID                            732-94-0499
Branch                                          C
City                                    Naypyitaw
Customer type                              Normal
Gender                                     Female
Product line               Electronic accessories
Unit price                                  41.65
Quantity                                       10
Tax 5%                                     20.825
Total                                     437.325
Date                                    1/13/2019
Time                                        17:04
Payment                               Credit card
cogs                                        416.5
gross margin percentage                  4.761905
gross income                               20.825
Rating                                        5.4
Name: 75, dtype: object

Рядок з максимальною ціною:
Invoice ID                            817-69-8206
Branch         

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [15]:
def categorize_rating(rating):
    if rating >= 8:
        return 'High'
    elif rating >= 6:
        return 'Medium'
    else:
        return 'Low'

df['Rating Category'] = df['Rating'].apply(categorize_rating)

print(df[['Rating', 'Rating Category']].head(10))

   Rating Rating Category
0     9.1            High
1     9.6            High
2     7.4          Medium
3     8.4            High
4     5.3             Low
5     4.1             Low
6     5.8             Low
7     8.0            High
8     7.2          Medium
9     5.9             Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [16]:
rating_percentage = df['Rating Category'].value_counts(normalize=True) * 100
print(rating_percentage)

Rating Category
Medium    35.6
High      32.9
Low       31.5
Name: proportion, dtype: float64


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [17]:
time = '11:28'

In [19]:
def get_hour(time_str):
    # Розділяємо рядок за двокрапкою і беремо першу частину (години)
    hour_str = time_str.split(':')[0]
    return int(hour_str)

# Перевірка роботи функції
hour = get_hour(time)
print(hour)

11


4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [23]:
df['Hour'] = df['Time'].apply(get_hour)

print(df[['Time', 'Hour']].head())

    Time  Hour
0  13:08    13
1  10:29    10
2  13:23    13
3  20:33    20
4  10:37    10


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [24]:
hour_counts = df['Hour'].value_counts().sort_index()
print(hour_counts)

Hour
10    101
11     90
12     89
13    103
14     83
15    102
16     77
17     74
18     93
19    113
20     75
Name: count, dtype: int64


Найбільше покупок здійснюється о **19:00** (113 покупок), а також помітно високі показники о **13:00** (103) та **15:00** (102) годинах.

Це логічно, адже:

**13:00** — це час обідньої перерви, коли люди можуть швидко зайти до супермаркету.

**15:00** — кінець робочого дня для деяких, коли також можна зробити покупки.

 **19:00** — вечірній час після роботи, коли люди часто купують продукти чи інші товари для дому.

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [26]:
average_total_by_city = df.groupby('City')['Total'].mean()
print(average_total_by_city)

City
Mandalay     319.872506
Naypyitaw    337.099715
Yangon       312.354031
Name: Total, dtype: float64


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [31]:
stats_by_city = df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2)
print(stats_by_city)

           count    mean    min      max
City                                    
Mandalay     332  319.87  18.64  1022.49
Naypyitaw    328  337.10  10.68  1042.65
Yangon       340  312.35  12.69  1039.29


Найбільше покупок здійснено в **Yangon** (340), а найменше — в **Naypyitaw** (328), але різниця невелика.

Середній чек найвищий у **Naypyitaw** (337.10), дещо нижчий у **Mandalay** (319.87) і найнижчий у **Yangon** (312.35).

Мінімальні чеки приблизно однакові, коливаються від 10 до 19.

Максимальний чек також досить схожий у всіх містах — близько 1000.



### Рекомендації:

 У **Naypyitaw** варто зосередитись на утриманні клієнтів із високими чеками та стимулюванні повторних покупок.

 У **Yangon** можна спробувати збільшити середній чек, наприклад, пропонуючи акційні набори або крос-продажі.

 У **Mandalay** також можна працювати над збільшенням кількості покупок, наприклад, через маркетингові кампанії або розширення асортименту.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [33]:
df.to_csv('supermarket_sales_enriched.csv', index=False)

In [34]:
df.head()

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   
4  373-73-7910      A     Yangon        Normal    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   
4       Sports and travel       86.31         7  30.2085  634.3785   2/8/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  \
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1   
1  10:29         Cash   76.40                 4.761905        3.8200     9.6   
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4   
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4   
4  10:37      Ewallet  604.17                 4.761905       30.2085     5.3   

  Rating Category  Hour  
0            High    13  
1            High    10  
2          Medium    13  
3            High    20  
4             Low    10